Source/NSFontCollection.m

```diff
@@ -0,0 +1,533 @@
+      NSDirectoryEnumerator	*de;
+      NSFontCollection		*newCollection;
+
+      if (_availableFontCollections == nil)
+	{
+	  // Create the global array of font collections...
+	  _availableFontCollections = [[NSMutableArray alloc] init];
+	}
+      else
+	{
+	  [_availableFontCollections removeAllObjects];
+	}
+      
+      /*
+       * Load color lists found in standard paths into the array
+       * FIXME: Check exactly where in the directory tree we should scan.
+       */
+      e = [NSSearchPathForDirectoriesInDomains(NSLibraryDirectory,
+	NSAllDomainsMask, YES) objectEnumerator];
+
+      while ((dir = (NSString *)[e nextObject])) 
+	{
+	  BOOL flag;
+
+	  dir = [dir stringByAppendingPathComponent: @"FontCollections"];
+	  if (![fm fileExistsAtPath: dir isDirectory: &flag] || !flag)
+	    {
+	      // Only process existing directories
+	      continue;
+	    }
+
+	  de = [fm enumeratorAtPath: dir];
+	  while ((file = [de nextObject])) 
+	    {
+	      if ([[file pathExtension] isEqualToString: @"collection"])
+		{
+		  NSString	*name;
+		  name = [file stringByDeletingPathExtension];
+		  newCollection = [NSKeyedUnarchiver unarchiveObjectWithFile: [dir stringByAppendingPathComponent: file]];
+                  if (newCollection != nil)
+                    {
+                      [_availableFontCollections addObject: newCollection];
+                      RELEASE(newCollection);
+                    }
+                }
+	    }
+	}  
+      /*
+      if (defaultSystemFontCollection != nil)
+        {
+	  [_availableFontCollections addObject: defaultSystemFontCollection];
+	}
+      */
+      [_fontCollectionLock unlock];
+    }
+}
+
+/*
+ * Writing and Removing Files
+ */
+- (BOOL) _writeToFile: (NSString *)path
+{
+  NSFileManager *fm = [NSFileManager defaultManager];
+  NSString      *tmpPath;
+  BOOL          isDir;
+  BOOL          success;
+  BOOL          path_is_standard = YES;
+
+  /*
+   * We need to initialize before saving, to avoid the new file being 
+   * counted as a different collection thus making it appear twice
+   */
+  [NSFontCollection _loadAvailableFontCollections];
+
+  if (path == nil)
+    {
+      NSArray	*paths;
+
+      // FIXME the standard path for saving font collections
+      paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory,
+	NSUserDomainMask, YES);
+      if ([paths count] == 0)
+	{
+	  NSLog (@"Failed to find Library directory for user");
+	  return NO;	// No directory to save to.
+	}
+      path = [[paths objectAtIndex: 0]
+	stringByAppendingPathComponent: @"FontCollections"]; 
+      isDir = YES;
+    }
+  else
+    {
+      [fm fileExistsAtPath: path isDirectory: &isDir];
+    }
+
+  if (isDir)
+    {
+      ASSIGN (_fullFileName, [[path stringByAppendingPathComponent: _name] 
+        stringByAppendingPathExtension: @"collection"]);
+    }
+  else // it is a file
+    {
+      if ([[path pathExtension] isEqual: @"collection"] == YES)
```

Comment:
The processing here and in the next if block are only needed when path has been explicitly specified, maybe it would be better to move that code up into that branch?